### ISSUE TITLE
[Docs] Question regarding SV-DKL paper/example / one GP per feature

### ISSUE THREAD

**Issue by cherepanovic** [cornellius-gp/gpytorch#947]
from [1]

> In the next cell, we create the layer of Gaussian process models that are called after the neural network. In this case, we'll be using one GP per feature, as in the SV-DKL paper. The outputs of these Gaussian processes will the be mixed in the softmax likelihood.

I am asking about the point **"we'll be using one GP per feature"**, where is this setting in the GaussianProcessLayer exactly? 

[1] - https://github.com/cornellius-gp/gpytorch/blob/master/examples/08_Deep_Kernel_Learning/Deep_Kernel_Learning_DenseNet_CIFAR_Tutorial.ipynb

**Comment by mmirtcho** [comment]
In the variational distribution, I think its the ``` batch_size=num_dim ``` and in the variational strategy it is the ``` num_dim=num_dim ```, I am like 90% certain.

**Comment by mmirtcho** [comment]
Does the CIFAR example work for you? I get this error: https://github.com/cornellius-gp/gpytorch/issues/945

**Comment by cherepanovic** [comment]
@mmirtcho  

> In the variational distribution, I think its the batch_size=num_dim and in the variational strategy it is the num_dim=num_dim, I am like 90% certain.

No, num_dim in variational_distribution is the number of features from NN, namely 784.

CIFAR example does work (wait please until your post will be answered).

**Comment by gpleiss** [comment]
That's happening through `num_dim=num_dim` right now. However, this example is out-of-date after the updates from #903. It will be updated to the new interface shortly.

**Comment by cherepanovic** [comment]
@gpleiss 

I thought, that  `num_dim` (in this example it is 784) is the dimension of g(x, w), loosely speaking, the dimension of the data samples which is the input for GP and which by the cov matrix will be built by RBF Kernel. Now, I am confused...

![1](https://user-images.githubusercontent.com/10064548/68983603-0da88880-080c-11ea-8c9e-61b1a390d59a.png)
![2](https://user-images.githubusercontent.com/10064548/68983606-11d4a600-080c-11ea-8eb6-401612d2a021.png)
source: DKL paper

I appreciate your help

**Comment by mmirtcho** [comment]
> @gpleiss
> 
> I thought, that `num_dim` (in this example it is 784) is the dimension of g(x, w), loosely speaking, the dimension of the data samples which is the input for GP and which by the cov matrix will be built by RBF Kernel. Now, I am confused... I rather thought in this example we have 10 GPs, each for every class...
> 
> ![1](https://user-images.githubusercontent.com/10064548/68983603-0da88880-080c-11ea-8c9e-61b1a390d59a.png)
> ![2](https://user-images.githubusercontent.com/10064548/68983606-11d4a600-080c-11ea-8eb6-401612d2a021.png)
> source: DKL paper
> 
> I appreciate your help

I think the whole point is that you assign the output of the DNN or DenseNet to the amount of classes so that at the end features=classes when passed to the GP. That is my understanding at least.

**Comment by cherepanovic** [comment]
> I think the whole point is that you assign the output of the DNN or DenseNet to the amount of classes so that at the end features=classes when passed to the GP. That is my understanding at least.

>>you assign the output of the DNN or DenseNet to the amount of classes

where do you see something in the code what refers to these 10 classes in case of CIFAR10?

**Comment by gpleiss** [comment]
> I thought, that num_dim (in this example it is 784) is the dimension of g(x, w), loosely speaking, the dimension of the data samples which is the input for GP and which by the cov matrix will be built by RBF Kernel.

No, `num_dim` is the number of dimensions output by the neural network (feature extractor). I.e. the neural network's output is `n x d`, where `n` is the number of data points in the minibatch, and `d = num_dim` is the number of features.

The `GPModel` learns an independent 1-dimensional GP for each of these `num_dim` features. So the output of `GPModel` is a 784-dimensional function.

The SoftmaxLikelihood reduces this 784-dimensional function into a 10-dimensional function through a linear mixing layer.

`likelihood = gpytorch.likelihoods.SoftmaxLikelihood(num_features=model.num_dim, num_classes=num_classes).cuda()`
